src/algebra/free.lean

```diff
@@ -2,173 +2,323 @@
 Copyright (c) 2019 Kenny Lau. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Kenny Lau
-
-1. Free magma of a type (traversable, decidable equality).
-2. Free semigroup of a magma.
-3. Free semigroup of a type (traversable, decidable equality).
-
-And finally, magma.free_semigroup (free_magma α) ≃ free_semigroup α.
 -/
 import data.equiv.basic
 import control.applicative
 
-universes u v
+/-!
+# Free constructions
+
+## Main definitions
+
+* `free_magma α`: free magma over alphabet `α`, defined inductively, with traversable instance
```

Comment:
Could you recall what a free magma is? To me, magma is a completely alien notion...

Reply:
done